File: ui/js/text_ui.js
/*
----------------------------------------
RIG Interactions (built for RIG-FiveM)

Author: Case (https://caseirl.dev)
Repo: https://github.com/rig-fivem/rig_interactions
License: https://github.com/rig-fivem/rig_interactions/blob/main/LICENSE
----------------------------------------
*/

export class TextUI {
    constructor(default_position = "bottom-center") {
        this.current_item = null;
        this.position = default_position;
    }

    resolve_position_style(position) {
        const map = {
            "top-left": "top:4vh;left:2vw;",
            "top-center": "top:4vh;left:50%;transform:translateX(-50%);",
            "top-right": "top:4vh;right:2vw;",
            "left-center": "top:50%;left:2vw;transform:translateY(-50%);",
            "right-center": "top:50%;right:2vw;transform:translateY(-50%);",
            "bottom-left": "bottom:4vh;left:2vw;",
            "bottom-center": "bottom:8vh;left:50%;transform:translateX(-50%);",
            "bottom-right": "bottom:4vh;right:2vw;"
        };
        return map[position] || map["bottom-center"];
    }

    set_data(data) {
        const next_position = data.position || this.position;
        const position_changed = next_position !== this.position;
        this.position = next_position;

        if (!$(".text_ui").length || position_changed) {
            this.destroy();
            this.build(data.image !== undefined);
        }

        this.current_item = data;
        this.update_display();
    }

    build(show_image) {
        const style = this.resolve_position_style(this.position);

        const content = `
            <div class="text_ui" style="${style}">
                ${show_image ? `<div class="text_ui_item"><img id="hint_image" src="" alt="Item Image"></div>` : ""}
                <div class="text_ui_message">
                    <p id="text_ui_status">NO ITEM EQUIPPED</p>
                    <p id="text_ui_action">Press E to Equip Item</p>
                </div>
            </div>
        `;
        $("#ui_focus").append(content);
    }

    update_display() {
        if (!this.current_item) {
            $("#text_ui_status").text("No Item Equipped");
            $("#text_ui_action").text("Press E to equip item");
            if ($("#hint_image").length) $("#hint_image").attr("src", "/ui/assets/images/no_image.png");
            return;
        }

        if (this.current_item.image && $("#hint_image").length) {
            $("#hint_image").attr("src", this.current_item.image);
        }

        if (this.current_item.quantity !== undefined) {
            $("#text_ui_status").html(`<div><span>${this.current_item.label || ""}</span><span id="hint_quantity">${this.current_item.quantity}x</span></div>`);
        } else {
            $("#text_ui_status").text(this.current_item.status_text || this.current_item.label || "");
        }

        $("#text_ui_action").text(this.current_item.action_text || "Press F to interact");
    }

    update_quantity(amount) {
        if (!this.current_item) return;
        this.current_item.quantity = amount;
        $("#hint_quantity").text(amount + "x");
        if (amount === 0) this.clear();
    }

    update_status_text(text) {
        $("#text_ui_status").text(text);
    }

    clear() {
        this.current_item = null;
        this.update_display();
    }

    destroy() {
        $(".text_ui").remove();
    }
}

File: src/client/systems/text_ui.lua
--[[
----------------------------------------
RIG Interactions (built for RIG-FiveM)

Author: Case (https://caseirl.dev)
Repo: https://github.com/rig-fivem/rig_interactions
License: https://github.com/rig-fivem/rig_interactions/blob/main/LICENSE
----------------------------------------
]]

--- @script text_ui
--- @file src/server/systems/text_ui.lua
--- @description Text UI display with image support

--- @section Functions

local function update_text_ui(data)
    SendNUIMessage({
        func = "update_text_ui",
        payload = data
    })
end

local function update_text_ui_quantity(amount)
    SendNUIMessage({
        func = "update_text_ui_quantity",
        payload = { amount = amount }
    })
end

local function clear_text_ui()
    SendNUIMessage({ func = "clear_text_ui" })
end

local function destroy_text_ui()
    SendNUIMessage({ func = "destroy_text_ui" })
end

--- @section Events

RegisterNetEvent("rig_interactions:client:update_text_ui", function(data)
    if not data then return print("data missing") end
    update_text_ui(data)
end)

RegisterNetEvent("rig_interactions:client:update_text_ui_quantity", function(amount)
    if not amount then return print("amount missing") end
    update_text_ui_quantity(amount)
end)

RegisterNetEvent("rig_interactions:client:clear_text_ui", function()
    clear_text_ui()
end)

RegisterNetEvent("rig_interactions:client:destroy_text_ui", function()
    destroy_text_ui()
end)

--- @section Exports

exports("update_text_ui", update_text_ui)
exports("update_text_ui_quantity", update_text_ui_quantity)
exports("clear_text_ui", clear_text_ui)
exports("destroy_text_ui", destroy_text_ui)

--- @section Test Commands

RegisterCommand("test_text_ui", function()
    print("testing text_ui")
    update_text_ui({
        image = "/ui/assets/images/RIG256.png",
        label = "Test Text UI",
        action_text = "Press F to do a thing"
    })
end)

RegisterCommand("test_text_ui_update", function()
    update_text_ui_quantity(2)
end)

RegisterCommand("test_text_ui_clear", function()
    clear_text_ui()
end)

RegisterCommand("test_text_ui_destroy", function()
    destroy_text_ui()
end)